onadata/libs/utils/hash.py

```diff
@@ -0,0 +1,175 @@
+
+import requests
+from django.conf import settings
+from rest_framework import status
+
+
+def get_hash(source: Union[str, bytes, BinaryIO],
+             algorithm: str = 'md5',
+             prefix: bool = False,
+             fast: bool = False) -> str:
+    """
+    Calculates the hash for an object.
+
+    :param source: string, bytes or FileObject. Files must be opened in binary mode  # noqa
+    :param algorithm: Can be 'md5' or 'sha1'. Default: 'md5'.
+    :param prefix: Prefix the return value with the algorithm, e.g.: 'md5:34523'
+    :param fast: If True, only calculate on 3 small pieces of the object.
+                 Useful with big (remote) files.
+    """
+
+    supported_algorithm = ['md5', 'sha1']
+
+    if algorithm not in supported_algorithm:
+        raise NotImplementedError('Only `{algorithms}` are supported'.format(
+            algorithms=', '.join(supported_algorithm)
+        ))
+
+    if algorithm == 'md5':
+        hashlib_def = hashlib.md5
+    else:
+        hashlib_def = hashlib.sha1
+
+    def _prefix_hash(hex_digest: str) -> str:
+        if prefix:
+            return f'{algorithm}:{hex_digest}'
+        return hex_digest
+
+    # If `source` is a string, it can be a URL or real string
+    if isinstance(source, str):
+        if not source.startswith('http'):
+            hashable = source.encode()
+        else:
+            # Try to get hash from the content of a URL.
+            # If the any requests fail, it returns the hash of the URL itself.
+            # We do not want to raise a `RequestException` while calculating
+            # the hash.
+            # ToDo: Evaluate whether it could be better to return ̀`None` when
+            # a request exception occurs
+
+            # Ensure we do not receive a gzip response.
+            headers = {'Accept-Encoding': None}
+            url_hash = hashlib_def(source.encode()).hexdigest()
+            # Get remote file size
+            # `requests.get(url, stream=True)` only gets headers. Body is
+            # only retrieved when `response.content` is called.
+            # It avoids making a second request to get the body
+            # (i.e.: vs `requests.head()`).
+            try:
+                response = requests.get(source, stream=True, headers=headers)
+                response.raise_for_status()
+            except requests.exceptions.RequestException:
+                # With `stream=True`, the connection is kept alive until it is
+                # closed or all the data is consumed. Because, we do not consume
+                # the data, we have to close the connexion manually.
+                response.close()
+                return _prefix_hash(hex_digest=url_hash)
+
+            file_size = int(response.headers['Content-Length'])
+            # If the remote file is smaller than the threshold or smaller than
+            # 3 times the chunk, we retrieve the whole file to avoid extra
+            # requests.
```

Comment:
Let's note in the `{algorithm}` when we use this approach

Reply:
Would you mind to remind me what you meant exactly? 
Do you want to add this in the docstrings ?